From: misakazip <git@misaka.zip>
Subject: [PATCH] Handle unavailable SecurityCenter CTA activities

Check whether either MIUI SecurityCenter CTA activity can be resolved.
When neither is available, skip the CTA flow and report it as unnecessary.
Keep Android runtime permission checks active.

diff --git a/smali_classes3/Sc/a.smali b/smali_classes3/Sc/a.smali
index 12e3e81..a84c39d 100644
--- a/smali_classes3/Sc/a.smali
+++ b/smali_classes3/Sc/a.smali
@@ -31,6 +31,18 @@
 .method public final b(Landroidx/fragment/app/l;Z)Landroid/content/Intent;
     .locals 5
 
+    invoke-static {}, LSc/a;->hasSecurityCenter()Z
+
+    move-result v0
+
+    if-nez v0, :cta_available
+
+    const/4 v0, 0x0
+
+    return-object v0
+
+    :cta_available
+
     new-instance p0, Landroid/content/Intent;
 
     const-string v0, "miui.intent.action.SYSTEM_PERMISSION_DECLARE_NEW"
@@ -220,6 +232,18 @@
 .method public final c()Z
     .locals 2
 
+    invoke-static {}, LSc/a;->hasSecurityCenter()Z
+
+    move-result v0
+
+    if-nez v0, :cta_available
+
+    const/4 v0, 0x1
+
+    return v0
+
+    :cta_available
+
     invoke-static {}, LZc/d;->a()LZc/a;
 
     move-result-object p0
@@ -252,3 +276,54 @@
 
     return p0
 .end method
+
+.method private static hasSecurityCenter()Z
+    .locals 3
+
+    invoke-static {}, Lcom/xiaomi/camera/basic/Global;->getContext()Landroid/content/Context;
+
+    move-result-object v0
+
+    invoke-virtual {v0}, Landroid/content/Context;->getPackageManager()Landroid/content/pm/PackageManager;
+
+    move-result-object v0
+
+    new-instance v1, Landroid/content/Intent;
+
+    const-string v2, "miui.intent.action.SYSTEM_PERMISSION_DECLARE_NEW"
+
+    invoke-direct {v1, v2}, Landroid/content/Intent;-><init>(Ljava/lang/String;)V
+
+    const-string v2, "com.miui.securitycenter"
+
+    invoke-virtual {v1, v2}, Landroid/content/Intent;->setPackage(Ljava/lang/String;)Landroid/content/Intent;
+
+    const/4 v2, 0x0
+
+    invoke-virtual {v0, v1, v2}, Landroid/content/pm/PackageManager;->resolveActivity(Landroid/content/Intent;I)Landroid/content/pm/ResolveInfo;
+
+    move-result-object v2
+
+    if-nez v2, :available
+
+    const-string v2, "miui.intent.action.SYSTEM_PERMISSION_DECLARE"
+
+    invoke-virtual {v1, v2}, Landroid/content/Intent;->setAction(Ljava/lang/String;)Landroid/content/Intent;
+
+    const/4 v2, 0x0
+
+    invoke-virtual {v0, v1, v2}, Landroid/content/pm/PackageManager;->resolveActivity(Landroid/content/Intent;I)Landroid/content/pm/ResolveInfo;
+
+    move-result-object v2
+
+    if-nez v2, :available
+
+    const/4 v0, 0x0
+
+    return v0
+
+    :available
+    const/4 v0, 0x1
+
+    return v0
+.end method
